### PR TITLE
[SPARK-28899][SQL][TEST] merge the testing in-memory v2 catalogs from catalyst and core

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

There are 2 in-memory `TableCatalog` and `Table` implementations for testing, in sql/catalyst and sql/core. This PR merges them.

After merging, there are 3 classes:
1. `InMemoryTable`
2. `InMemoryTableCatalog`
3. `StagingInMemoryTableCatalog`

For better maintainability, these 3 classes are put in 3 different files.

### Why are the changes needed?

reduce duplicated code

### Does this PR introduce any user-facing change?

no
### How was this patch tested?

N/A

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTableCatalog.scala [hunk 1]
```diff
@@ -63,35 +65,42 @@ class TestTableCatalog extends TableCatalog with SupportsNamespaces {
       schema: StructType,
       partitions: Array[Transform],
       properties: util.Map[String, String]): Table = {
-
     if (tables.containsKey(ident)) {
       throw new TableAlreadyExistsException(ident)
     }
 
-    if (partitions.nonEmpty) {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryTableCatalog.scala]: the sql/core version supports partitions, so the merged version should support partitions as well.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/TestInMemoryTableCatalog.scala [hunk 2]
```diff
@@ -1,535 +0,0 @@
...
-import org.apache.spark.sql.catalyst.InternalRow
-import org.apache.spark.sql.catalyst.analysis.{CannotReplaceMissingTableException, NamespaceAlreadyExistsException, NoSuchNamespaceException, NoSuchTableException, TableAlreadyExistsException}
-import org.apache.spark.sql.sources.{And, EqualTo, Filter, IsNotNull}
-import org.apache.spark.sql.sources.v2.reader.{Batch, InputPartition, PartitionReader, PartitionReaderFactory, Scan, ScanBuilder}
-import org.apache.spark.sql.sources.v2.writer.{BatchWrite, DataWriter, DataWriterFactory, SupportsDynamicOverwrite, SupportsOverwrite, SupportsTruncate, WriteBuilder, WriterCommitMessage}
-import org.apache.spark.sql.types.StructType
-import org.apache.spark.sql.util.CaseInsensitiveStringMap
-
-// this is currently in the spark-sql module because the read and write API is not in catalyst
-// TODO(rdblue): when the v2 source API is in catalyst, merge with TestTableCatalog/InMemoryTable
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/TestInMemoryTableCatalog.scala]: This TODO is completed in this PR.

**Comment by cloud-fan** [comment]
cc @rdblue @brkyvz

**Comment by rdblue** [comment]
+1

**Comment by rdblue** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
Yey! Congratulation, @rdblue !!! 😄

**Comment by rdblue** [comment]
Thanks!

**Comment by cloud-fan** [comment]
@rdblue congrats and thanks for merging my PR! Hopefully this is your first Spark PR merging :)

**Pull request closed without merging**